pkg/pillar/pubsub/driver.go

```diff
@@ -0,0 +1,58 @@
+package pubsub
+
+// Driver a backend driver for pubsub
+type Driver interface {
+	// Publisher return a `DriverPublisher` for the given name and topic.
+	// The caller passes the `Updaters`, `Restarted` checker and `Differ`.
+	// These can be used to:
+	// * add to or remove from the updaters
+	// * determine if the topic has been restarted
+	// * diff the current known state from the target known state
+	Publisher(global bool, name, topic string, persistent bool, updaterList *Updaters, restarted Restarted, differ Differ) (DriverPublisher, error)
+	// Subscriber return a `DriverSubscriber` for the given name and topic.
+	// This is expected to create a `DriverSubscriber`, but not start it.
+	// Once started, when changes arrive, they should be published to the provided
+	// channel. Each update to the channel is of type `Change`, which encapsulates
+	// the operation, key and value.
+	Subscriber(global bool, name, topic string, persistent bool, C chan Change) (DriverSubscriber, error)
+}
+
+// DriverSubscriber interface that a driver for subscribing must implement
+type DriverSubscriber interface {
+	// Start subscribing to a name and topic and publish changes to the channel
+	Start() error
+	// DefaultName Return the default name to use for an agent, when the name
+	// is not provided.
+	DefaultName() string
```

Comment:
This comment doesn't quite explain it to me, and DefaultName isn't very clear. However, I don't have a good term for it in general. Perhaps this is the ExternalName i.e., the name to use to subscribe something which isn't published by pubsub but published in the form of a (json) file. But that assumes a particular implementation of the driver.